File: codility/lessons/7_stack_and_queue/fish.py
# https://app.codility.com/programmers/lessons/7-stacks_and_queues/fish/

UPSTREAM, DOWNSTREAM = 0, 1

def solution(A, B):

    fishes_alive = len(A)
    downstream_stack = []

    for (fish_size, direction) in zip(A, B):

        if direction == DOWNSTREAM:
            downstream_stack.append(fish_size)
            continue

        while downstream_stack and fish_size > downstream_stack[-1]:
            fishes_alive -= 1
            downstream_stack.pop()

        if downstream_stack and fish_size < downstream_stack[-1]:
            fishes_alive -= 1

    return fishes_alive

File: codility/lessons/7_stack_and_queue/fish.cpp
#include <stack>
#include <vector>
using namespace std;

enum class Direction : int {Downstream = 1, Upstream = 0};

int solution(vector<int> &ranks, vector<int> &directions) {
    int alive = 0;
    stack<int> downstream_ranks;

    for (size_t i = 0; i < ranks.size(); i++) {
        if (static_cast<Direction>(directions[i]) == Direction::Downstream) {
            downstream_ranks.push(ranks[i]);
            alive++;
        }
        if (static_cast<Direction>(directions[i]) == Direction::Upstream) {
            while (!downstream_ranks.empty() && downstream_ranks.top() < ranks[i]) {
                downstream_ranks.pop();
                alive--;
            }
            if (downstream_ranks.empty()) {
                alive++;
            }
        }
    }

    return alive;
}
